### ISSUE TITLE
UnnecessarySemicolonRule throws exceptions for Groovy 3

### ISSUE THREAD

**Issue by dmurat** [CodeNarc/CodeNarc#506]
I'm trying to run the latest CodeNarc from `master` on a Groovy 3 project (an exact groovy version is 3.0.4). However, UneccessarySemicoloRule throws IndexOutOfBoundException on line 69. After changing the `checkLastLineForSemicolon` method as shown below, everything started to work correctly.

    private void checkLastLineForSemicolon(SourceCode sourceCode, List<Violation> violations, AnnotatedNode node) {
        int lineNumber = node.getLastLineNumber()
        if (lineNumber < 1) { // <--- check added
            return
        }

        String line = sourceCode.getLines().get(lineNumber - 1) + ' '   // to make it easier to extract the final chars
        int lastColumn = node.lastColumnNumber
        boolean lastCharIsSemicolon = line[lastColumn - 1] == ';'

        if (lastCharIsSemicolon) {
            violations << createViolation(lineNumber, line, MESSAGE)
        }
    }

Tnx

**Comment by dmurat** [comment]
I believe the root cause is in Groovy - https://issues.apache.org/jira/browse/GROOVY-9577
I'm not sure if it makes sense fixing it in CondeNarc. Probably not.

**Comment by chrismair** [comment]
Can you please provide some sample code that causes the above exception, so I can add a test for that.

**Comment by dmurat** [comment]
Sure, there is a test that triggered the problem for me:

    @Test
    void test_NoViolations_WithClassAnnotations() {
        final SOURCE = '''
            package foo
            
            import java.io.CharArrayReader
            import java.nio.file.AccessMode
            import java.io.Bits
            
            import java.lang.String

            class A {
                String firstName
                
                int a() {
                    return 1
                }
            }
        '''

        assertNoViolations(SOURCE)
    }

I added the test in `UnnecessarySemicolonRuleTest` and run it with following changes in `build.gradle`

    def groovyTestVersion = '3.0.4'
    ...
    dependencies {
      ...
      testRuntimeOnly("org.codehaus.groovy:groovy:$groovyTestVersion")
      testRuntimeOnly("org.codehaus.groovy:groovy-xml:$groovyTestVersion")
      testRuntimeOnly("org.codehaus.groovy:groovy-json:$groovyTestVersion")
      testRuntimeOnly("org.codehaus.groovy:groovy-ant:$groovyTestVersion")
      testRuntimeOnly("org.codehaus.groovy:groovy-templates:$groovyTestVersion")
    }

Basically, whenever you add some package that Groovy does not include by default, the error can occur. If it passes, try adding more imports from "non-standard" packages.

**Comment by dmurat** [comment]
I just noticed that the name of a test is wrong. Previously I suspected that rule failed because of additional annotations on a class. But it turns out that imports are the problem.

**Comment by chrismair** [comment]
Thanks for providing the sample code and pointing out the fix!
